controllers/keda/prometheus_metrics_controller.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2022 The KEDA Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keda
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	"github.com/kedacore/keda/v2/pkg/metrics"
+)
+
+type PrometheusMetricsReconciler struct {
```

Comment:
Do we really need a separate Reconciller for this? Can't this calculation be done in the existing ScaledObject/ScaledJob controller? or in scale_handler?

Reply:
The reconcile methods in the scaled object / job controllers are passed in only the new object. To calculate the number of active triggers accurately, we need the previous spec of the scaled object as well, to compute the diff. 

The only way then to calculate this is on every reconcile fetch all scaled object / jobs and use them to calculate the totals. This would have to be done in both the controllers. Having to fetch all the scaled objects when a scaled job is being reconciled in the scaled job controller or vice versa, was not that clean to me.

I thought it was cleaner to have a separate reconcile loop which lists both of them and updates the metric.

Reply:
In the future, if there are more metrics they could also be possibly added to this controller, I feel.

Reply:
> The reconcile methods in the scaled object / job controllers are passed in only the new object. To calculate the number of active triggers accurately, we need the previous spec of the scaled object as well, to compute the diff.
> 

The reconcile loop is also executed on updated/removed ScaledObject/Job, so I think there's a way how to include it to the current stuf.

For example here is handled the calculation of all `metricNames` that's provided by all ScaledObjects in the cluster. there could be the same `metricName` for multiple SOs. This might give you some inspiration.
https://github.com/kedacore/keda/blob/main/controllers/keda/metrics_adapter_controller.go#L107

> The only way then to calculate this is on every reconcile fetch all scaled object / jobs and use them to calculate the totals. This would have to be done in both the controllers. Having to fetch all the scaled objects when a scaled job is being reconciled in the scaled job controller or vice versa, was not that clean to me.
> 

Imagine having 1000s of ScaledObjects and with each change trigger a listing of all of them. I know that it should hit the cache, but anyway, I am not sure this is opti

> I thought it was cleaner to have a separate reconcile loop which lists both of them and updates the metric.

I'd really avoid introducing a new reconiler querying for all SOs/SJs in the cluster with each update.

Reply:
I'LL take a look.

Reply:
What do you mean when you say include it to the current stuff?

Reply:
I have been thinking about this and maybe we could create a `map[string]string` where the key is SO/JS name+namespace. On every ScaledObject/Job reconciliation loop we could update the map with the current value and remove it with the finalizers. Once we have this, the Prometheus metric will evaluate that map on every request but instead of listing the objects from the cluster, they are in memory. WDYT?

Reply:
This is what I am working on after looking at the code pointer commented by @zroubalik . Almost done testing changes, will raise the PR soon.